codesign/codesign.go

```diff
@@ -0,0 +1,338 @@
+	}
+
+	authConfig, err := appleauth.Select(conn, []appleauth.Source{authSource}, appleauth.Inputs{})
+	if err != nil {
+		if conn != nil && conn.APIKeyConnection == nil && conn.AppleIDConnection == nil {
+			fmt.Println()
+			logger.Warnf("%s", devportalclient.NotConnectedWarning)
+		}
+
+		if errors.Is(err, &appleauth.MissingAuthConfigError{}) {
+			if authType == AppleIDAuth {
+				return appleauth.Credentials{}, fmt.Errorf("Apple ID authentication is selected in Step inputs, but Bitrise Apple service connection is unset")
+			}
+
+			return appleauth.Credentials{}, fmt.Errorf("API key authentication is selected in Step inputs, but Bitrise Apple service connection is unset")
+		}
+
+		return appleauth.Credentials{}, fmt.Errorf("could not select Apple authentication credentials: %w", err)
+	}
+
+	if authConfig.APIKey != nil {
+		authConfig.AppleID = nil
+
+		logger.Donef("Using Apple service connection with API key.")
+	} else if authConfig.AppleID != nil {
+		logger.Donef("Using Apple service connection with Apple ID.")
+	} else {
+		panic("No Apple authentication credentials found.")
+	}
+
+	return authConfig, nil
+}
+
+func (m *Manager) selectCodeSigningStrategy(credentials appleauth.Credentials) (codeSigningStrategy, string, error) {
+	const manualProfilesReason = "Using Bitrise-managed code-signing via API key, as Automatically managed signing is disabled in Xcode for the project."
+
+	if credentials.AppleID != nil {
+		return codeSigningBitriseAppleID, "Using Bitrise-managed code-signing via Apple ID, as Apple ID is not supported by Xcode-managed code-signing.", nil
+	}
+
+	if credentials.APIKey == nil {
+		panic("No Apple authentication credentials found.")
+	}
+
+	if !m.opts.ShouldConsiderXcodeSigning {
+		return codeSigningBitriseAPIKey, "", nil
+	}
+
+	if m.opts.XcodeMajorVersion < 13 {
+		return codeSigningBitriseAPIKey, "Using Bitrise-managed code-signing via API key, as Xcode-managed code-signing requires at least Xcode 13.", nil
+	}
+
+	project, err := m.getProject()
+	if err != nil {
+		return codeSigningXcode, "Using Xcode-managed code-signing, as project parsing failed.", err
+	}
+
+	isManaged, err := project.IsSigningManagedAutomatically()
+	if err != nil {
+		return codeSigningBitriseAPIKey, manualProfilesReason, err
+	}
+
+	if isManaged {
+		return codeSigningXcode, "Using Xcode-managed code-signing, as Automatically managed signing is enabled in Xcode for the project", nil
+	}
+
+	return codeSigningBitriseAPIKey, manualProfilesReason, nil
+}
+
+func (m *Manager) downloadAndInstallCertificates() error {
+	certificates, err := m.certDownloader.GetCertificates()
+	if err != nil {
+		return fmt.Errorf("failed to download certificates: %s", err)
+	}
+
+	m.logger.Infof("Installing downloaded certificates:")
+	for _, cert := range certificates {
+		m.logger.Printf("- %s", cert)
+		// Empty passphrase provided, as already parsed certificate + private key
+		if err := m.keychain.InstallCertificate(cert, ""); err != nil {
```

Comment:
in case of Bitrise managed auto codesigning we install a single certificate per distribution type, should not we do so here as well?

Reply:
Would not select a single one as Xcode can select one later, but at least validate that development cert is avaialble (if needed)